web/jsonapi/errors.go

```diff
@@ -0,0 +1,91 @@
+package jsonapi
+
+import (
+	"net/http"
+
+	"github.com/cozy/cozy-stack/vfs"
+)
+
+// SourceError contains references to the source of the error
+type SourceError struct {
+	Pointer   string `json:"pointer,omitempty"`
+	Parameter string `json:"parameter,omitempty"`
+}
+
+// Error objects provide additional information about problems encountered
+// while performing an operation.
+// See http://jsonapi.org/format/#error-objects
+type Error struct {
+	Status int         `json:"status,string"`
+	Title  string      `json:"title"`
+	Detail string      `json:"detail"`
+	Source SourceError `json:"source,omitempty"`
+}
+
+// WrapVfsError returns a formatted error from a golang error emitted by the vfs
+func WrapVfsError(err error) *Error {
+	switch err {
+	case vfs.ErrDocAlreadyExists:
+		return &Error{
+			Status: http.StatusConflict,
+			Title:  "Conflict",
+			Detail: err.Error(),
+		}
+	case vfs.ErrDocDoesNotExist:
+		return NotFound(err)
+	case vfs.ErrParentDoesNotExist:
+		return NotFound(err)
+	case vfs.ErrDocTypeInvalid:
+		return InvalidAttribute("type", err)
+	case vfs.ErrIllegalFilename:
+		return InvalidParameter("folder-id", err)
+	case vfs.ErrInvalidHash:
+		return InvalidParameter("Content-MD5", err)
```

Comment:
Same remark for `ErrContentLengthMismatch` which can fall into the same category.